shiny/_template_utils.py

```diff
@@ -0,0 +1,166 @@
+import shutil
+import sys
+from pathlib import Path
+from typing import List, Optional, Tuple
+
+import questionary
+from questionary import Choice
+
+from ._custom_component_template_questions import (
+    install_js_dependencies,
+    isValidName,
+    updateComponentNameInTemplate,
+)
+
+
+def template_query(question_state: Optional[str] = "start"):
+    """
+    This will initiate a CLI query which will ask the user which template they would like.
+    If called without arguments this function will start from the top level and ask which
+    type of template the user would like.
+
+    You can also specify a question state to return to another level. For example if you
+    were at level 5 of a question chain and wanted to return to level 4.
+    This is not that useful currently because we only have two levels of questions.
+
+
+    :param question_state: The question state you would like to return to. Currently, the options are:
+        "cancel": Cancel the operation and exit.
+        "js-component": Start the questions for creating a custom JavaScript component.
+    """
+    if question_state is None:
+        template = questionary.select(
+            "Which template would you like to use?:",
+            choices=choicesArray(
+                [
+                    ("Basic App", "basic-app"),
+                    ("Express app", "express"),
+                    ("Dashboard", "dashboard"),
+                    ("Multi-page app with modules", "multi-page"),
+                    ("Custom JavaScript Component", "js-component"),
+                    ("Cancel", "cancel"),
+                ]
+            ),
+        ).ask()
+    else:
+        template = question_state
+
+    # Define the control flow for the top level menu
+    # This is simple now but will get more complicated as we add more templates
+    flow_dispatch = {
+        "cancel": lambda: sys.exit(),
+        "js-component": lambda: jsComponentQuestions(),
+        "default": lambda: appTemplateQuestions(template),
+    }
+
+    # Call the appropriate function based on the user's choice
+    flow_dispatch.get(template, flow_dispatch["default"])()
```

Comment:
This feels like unnecessary complexity even given the comment, but maybe I'm missing something about how we're expecting this to evolve.

```suggestion
    # Define the control flow for the top level menu
    if template is None or template == "cancel":
        sys.exit(1)
    elif template == "js-component":
        jsComponentQuestions()
        return
    else:
        appTemplateQuestions(template)
```

Reply:
Very fair. The problem of dispatching to a given action based on a string is one I always want to reach for a `switch` statement with but, as I hinted at in the comment: it's most likely overkill here. I think I had some idea of a future where the branches got complex and we used this dispatch object to control navigation and backing up etc based on a path of "keys" used to get to a given step.